### DISCUSSION TITLE
Consider ECVRF

### DISCUSSION THREAD

**Discussion by iquerejeta** [jedisct1/libsodium#1177] [Ideas]
Hello! I saw that in the 1.0.x roadmap, there is the point of `Consider ECVRF`. Is this still something that is being considered? I would be happy to do a first draft implementation compatible with the current version of the VRF draft. 

Otherwise, are you planning to expose a more efficient API for EC arithmetic? Current API prioritises security over efficiency (with which I agree), and forces users of the library to perform to/from bytes conversion, and checks (canonical, small order, main subgroup) for every arithmetic operation exposed in the API. This hampers the usage of the library for constructing this type of higher level protocols. 

Thank you!

**Comment by jedisct1** [comment]
VRFs are useful, and yes, it would be a good addition to libsodium.

Revision 11 of the draft was published and we're getting very close to it moving to an RFC. I'd still wait until this happens before merging relevant code, though, as last-minute breaking changes remain way too common in standardization processes. An implementation can still be prepared independently using the current draft.

The existing low-level APIs try to stay consistent with other APIs, by not introducing new representations for scalars and elements, and by trying to mitigate the implications of misuse. Duplicating them to add variants that do the same thing, but are easier to misuse, harder to write bindings for, and require inconsistent conventions would deviate too much from the original goal of this project. Performance is not a goal for libsodium.

That being said, the Ristretto255 encoding is much cheaper than the Edwards25519 one (validation only requires an inequality check), so maybe this is something you can consider for your protocols?

**Reply by iquerejeta** [reply]
That's great. I'll prepare a draft PR with revision 11, so we can start iterating over that. 

And the actual project I had in mind (that we are using) is ECVRF over Edwards25519, so we are currently constrained to that. I'm happy to hear that a VRF is considered to be a good addition to libsodium.

**Comment by iquerejeta** [comment]
I've implemented the latest version of the draft. 

https://github.com/iquerejeta/libsodium/tree/vrf_version_12

I tried to follow the style of ed25519, given the similarities between both primitives. It might even be worth considering to take some functions (key generation & seeded key generation) directly from ed25519, what do you think? I was thinking of waiting until standardisation is closed to open a PR (to avoid having a PR open for a while), but I'm happy to open it and have the discussion there if you prefer.

**Reply by jedisct1** [reply]
These changes look really good!

**Reply by iquerejeta** [reply]
Thanks! I'll open a PR 👍

**Comment by gitmachtl** [comment]
Please implement it, would be a very useful addition!

**Comment by Argimirodelpozo** [comment]
Hi everyone,
I wanted to revisit this old discussion around the VRF implementation.

A while back, a PR was opened by @iquerejeta to add it to the library, but it was eventually dropped due to the use of variable-length arrays.
I've now done [a small reworking of the implementation](https://github.com/Argimirodelpozo/libsodium/tree/vrf_impl) to remove VLAs entirely so it can be safely included in the main library. 

While doing this, I also added support for the earlier ietf-draft-03 version of the spec. The rationale behind this is that it’s widely used in the web3 ecosystem (both Cardano and Algorand implement it in their sodium forks). I kept it cleanly decoupled from the standardized RFC 9381 implementation, and the draft version can be easily dropped if there are objections to including it, but I think it’s worth having it available.

Would love to hear your thoughts on whether this could be reconsidered for inclusion. If so I can go ahead and open a PR.
Thank you!
